File: pulumi/folio/vpc.py
import pulumi
from pulumi_aws import ec2, get_availability_zones
from tags import standard_tags

## VPC

name = "folio-eks-vpc"
vpc = ec2.Vpc(
    name,
    cidr_block="10.100.0.0/16",
    instance_tenancy="default",
    enable_dns_hostnames=True,
    enable_dns_support=True,
    tags=standard_tags(name)
)

name = "folio-public-vpc-ig"
igw = ec2.InternetGateway(
    name,
    vpc_id=vpc.id,
    tags=standard_tags(name)
)

# If a subnet's traffic is routed to a internet gateway it is public.

name = "folio-public-vpc-route-table"
eks_route_table = ec2.RouteTable(
    name,
    vpc_id=vpc.id,
    routes=[
        ec2.RouteTableRouteArgs(
            cidr_block="0.0.0.0/0",
            gateway_id=igw.id,
        )
    ],
    tags=standard_tags(name)
)

## Subnets, one public and one private for each AZ in a region.

# For the definition of public and private subnets see:
# https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Subnets.html#vpc-subnet-basics

zones = get_availability_zones()
subnet_ids = []
private_subnet_ids = []

# Create the public subnets. Public subnets are public because they
# are associated with an InternetGateway.
for zone in zones.names:
    vpc_subnet = ec2.Subnet(
        f"folio-public-vpc-subnet-{zone}",
        assign_ipv6_address_on_creation=False,
        vpc_id=vpc.id,
        map_public_ip_on_launch=True,
        cidr_block=f"10.100.{len(subnet_ids)}.0/24",
        availability_zone=zone,
        tags=standard_tags(f"folio-public-sn-{zone}")
    )
    ec2.RouteTableAssociation(
        f"folio-public-vpc-route-table-assoc-{zone}",
        route_table_id=eks_route_table.id,
        subnet_id=vpc_subnet.id,
    )
    subnet_ids.append(vpc_subnet.id)

# Create the private subnets. Private subnets are private because
# they are NOT associated with an InternetGateway or other gateway.
for zone in zones.names:
    vpc_subnet = ec2.Subnet(
        f"folio-private-vpc-subnet-{zone}",
        assign_ipv6_address_on_creation=False,
        vpc_id=vpc.id,
        map_public_ip_on_launch=True,
        cidr_block=f"10.100.{len(subnet_ids)}.0/24",
        availability_zone=zone,
        tags=standard_tags(f"folio-private-sn-{zone}")
    )
    subnet_ids.append(vpc_subnet.id)
    private_subnet_ids.append(vpc_subnet.id)

## Security Group

name = "folio-eks-cluster-sg"
eks_security_group = ec2.SecurityGroup(
    name,
    vpc_id=vpc.id,
    description="Allow all HTTP(s) traffic to EKS Cluster",
    tags=standard_tags(name),
    ingress=[
        ec2.SecurityGroupIngressArgs(
            cidr_blocks=["0.0.0.0/0"],
            from_port=443,
            to_port=443,
            protocol="tcp",
            description="Allow pods to communicate with the cluster API Server.",
        ),
        ec2.SecurityGroupIngressArgs(
            cidr_blocks=["0.0.0.0/0"],
            from_port=80,
            to_port=80,
            protocol="tcp",
            description="Allow internet access to pods",
        ),
    ],
)

File: pulumi/folio/vpc.ts
import * as awsx from "@pulumi/awsx";
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import { Output } from "@pulumi/pulumi";

export module deploy {
    export function awsVpc(
        name: string, tags: object, availabilityZones: number, natGateways: number, cidrBlock: string): awsx.ec2.Vpc {
        const vpc = new awsx.ec2.Vpc(name, {

            tags: { "Name": name, ...tags },

            // If we have too many AZs and NAT gateways we'll run out of EIPs with our
            // current quota.
            numberOfAvailabilityZones: availabilityZones,

            // We could have these be equal to the number of availability zones for greater
            // fault tolerance. Although the cost will be higher.
            numberOfNatGateways: natGateways,

            // See https://github.com/pulumi/pulumi-eks/blob/e03751425604b5193c00cd2f8858d9b16f7901f0/examples/subnet-tags/index.ts
            // for what's going on with the tags here.
            subnets: [{
                type: "public", name: "folio-subnet",
                tags: { "kubernetes.io/role/elb": "1", ...tags },
            },
            {
                type: "private", name: "folio-subnet",
                tags: { "kubernetes.io/role/internal-elb": "1", ...tags }
            }],

            cidrBlock: cidrBlock
        },
            {
                // Inform pulumi to ignore tag changes to the VPCs or subnets, so that
                // tags auto-added by AWS EKS do not get removed during future
                // refreshes and updates, as they are added outside of pulumi's management
                // and would be removed otherwise.
                // Also see: https://github.com/pulumi/pulumi-eks/blob/master/examples/subnet-tags/index.ts
                // for links to issues with the background on this.
                transformations: [(args: any) => {
                    if (args.type === "aws:ec2/vpc:Vpc" ||
                        args.type === "aws:ec2/subnet:Subnet") {
                        return {
                            props: args.props,
                            opts: pulumi.mergeOptions(args.opts, { ignoreChanges: ["tags"] }),
                        };
                    }
                    return undefined;
                }],
            }
        );

        return vpc;
    }

    export function awsSecurityGroup
        (name: string,
         tags: object,
         vpcId: Output<string>,
         clusterCidrBlock: string): aws.ec2.SecurityGroup {
        // Create the security group. We need a custom security group since we're
        // going to expose non-standard ports for things like edge modules. This
        // is using the pulumi Classic API. Although Crosswalk (awsx) has its own
        // SecurityGroup type, Cluster wants a Classic SecurityGroup type.
        // This security group, being the security group for the VPC, also controls
        // access to our RDS cluster.
        const sg = new aws.ec2.SecurityGroup(name, {
            tags: { "Name": name, ...tags },
            vpcId: vpcId,
            // TODO I'm not loving this. These should be params I think.
            ingress: [{
                description: "Allow inbound traffic on 443",
                fromPort: 443,
                toPort: 443,
                protocol: "tcp",
                cidrBlocks: ["0.0.0.0/0"]
            }, {
                description: "Allow inbound traffic to okapi",
                fromPort: 9130,
                toPort: 9130,
                protocol: "tcp",
                cidrBlocks: ["0.0.0.0/0"]
            }, {
                description: "Allow inbound traffic on 9000 for edgeconexion",
                fromPort: 9000,
                toPort: 9000,
                protocol: "tcp",
                cidrBlocks: ["0.0.0.0/0"]
            }, {
                description: "Allow inbound traffic on 5432 for postgres",
                fromPort: 5432,
                toPort: 5432,
                protocol: "tcp",
                cidrBlocks: [clusterCidrBlock]
            },
            // TODO These can be removed as they are not in scope since we're deploying os out of vpc.
            {
                description: "Allow inbound traffic on 9200 for opensearch",
                fromPort: 9200,
                toPort: 9200,
                protocol: "tcp",
                cidrBlocks: [clusterCidrBlock]
            }, {
                description: "Allow inbound traffic to opensearch dashboard",
                fromPort: 5601,
                toPort: 5601,
                protocol: "tcp",
                cidrBlocks: [clusterCidrBlock]
            }],
            egress: [{
                description: "Allow outbound traffic",
                fromPort: 0,
                toPort: 0,
                protocol: "-1",
                cidrBlocks: ["0.0.0.0/0"]
            }]
        });

        return sg;
    }
}
